membership-attribute-service/app/controllers/AccountController.scala

```diff
@@ -236,24 +241,53 @@ class AccountController(commonActions: CommonActions, override val controllerCom
     }
   }
 
+  def checkForGiftSubscription(
+    userId: String,
+    subscriptionService: SubscriptionService[Future],
+    zuoraRestService: ZuoraRestService[Future],
+    nonGiftSubs: List[ContactAndSubscription],
+    contact: Contact
+  ): OptionT[FutureEither, List[ContactAndSubscription]] = {
+    val giftSub = for {
+      records <- OptionEither.liftOption(zuoraRestService.getGiftSubscriptionRecordsFromIdentityId(userId))
+      result <- if (records.isEmpty)
+        OptionEither.liftFutureEither[Subscription[AnyPlan]](None)
+      else
+        OptionEither.liftFutureOption(subscriptionService.get[AnyPlan](Name(records.head.Name), isActiveToday = true))
+    } yield result
+    val fullList = giftSub
+      .map(sub => ContactAndSubscription(contact, sub) :: nonGiftSubs)
+      .getOrElse(nonGiftSubs)
+    OptionEither.liftOption(fullList.run)
+  }
+
   def allCurrentSubscriptions(
     contactRepo: SimpleContactRepository,
-    subService: SubscriptionService[Future]
+    subService: SubscriptionService[Future],
+    zuoraRestService: ZuoraRestService[Future]
   )(
       maybeUserId: Option[String],
       filter: OptionalSubscriptionsFilter
   ): OptionT[OptionEither.FutureEither, List[ContactAndSubscription]] = for {
-      user <- OptionEither.liftFutureEither(maybeUserId)
-      contact <- OptionEither(contactRepo.get(user))
-      contactAndSubscriptions <-
+    userId <- OptionEither.liftFutureEither(maybeUserId)
+    contact <- OptionEither(contactRepo.get(userId))
+    nonGiftContactAndSubscriptions <-
         OptionEither.liftEitherOption(
           subService.current[SubscriptionPlan.AnyPlan](contact) map {
             _ map { subscription =>
               ContactAndSubscription(contact, subscription)
             }
           }
-        ) // TODO are we happy with an empty list in case of error ?!?!
-      filteredIfApplicable = filter match {
+        )
+
+    contactAndSubscriptions <- checkForGiftSubscription(
+        userId,
+        subService,
+        zuoraRestService,
+        nonGiftContactAndSubscriptions,
```

Comment:
not so keen on passing these in just to have them passed back.  is it worth returning an option and then doing something like  `List( nonGiftContactAndSubscriptions, giftContactAndSubscriptions).flatten` to make the higher level structure clearer?

Reply:
It's really tough to work through the types in the parent method, the comprehension is already using an `OptionT[FutureEither, A]` so we'd end up having to return a `OptionT[FutureEither, Option[Subscription]` somehow which is just horrendous. I've had a quick go at it and it doesn't seem to line up properly anyway which was why I decided to just encapsulate the whole mess inside one method and return the type we actually want.